x/shield/keeper/proposal.go

```diff
@@ -17,3 +21,339 @@ func (k Keeper) ClaimUnlock(ctx sdk.Context, proposalID, poolID uint64, loss sdk
+		//            delegations                     unbondings
+		// -------------------------------|---------------------------------
+		payoutFromDelegation = provider.DelegationBonded.Sub(purchased)
+		payoutFromUnbonding = payout.Sub(payoutFromDelegation)
+	} else {
+		// If delegation < purchased:
+		//                         |        withdraw      |
+		//                      purchased     | payout |
+		//                    -----|----------'--------'--|-----
+		//            delegations                     unbondings
+		// -------------------------------|---------------------------------
+		// or
+		//                                   |        withdraw      |
+		//                                purchased     | payout |
+		//                              -----|----------'--------'--|-----
+		//            delegations                     unbondings
+		// -------------------------------|---------------------------------
+		payoutFromUnbonding = payout
+	}
+
+	if payoutFromDelegation.IsPositive() {
+		k.PayFromDelegation(ctx, provider.Address, payoutFromDelegation)
+	}
+
+	if payoutFromUnbonding.IsPositive() {
+		uncoveredPurchase := sdk.MaxInt(sdk.ZeroInt(), purchased.Sub(provider.DelegationBonded))
+		unbondingDelegations := k.GetSortedUnbondingDelegations(ctx, provider.Address)
+		for _, ubd := range unbondingDelegations {
+			entry := ubd.Entries[0]
+			// If purchased is not fully covered, cover purchased first.
+			remainingUbd := entry.Balance
+			if uncoveredPurchase.IsPositive() {
+				if uncoveredPurchase.GTE(entry.Balance) {
+					uncoveredPurchase = uncoveredPurchase.Sub(entry.Balance)
+					remainingUbd = sdk.ZeroInt()
+				} else {
+					remainingUbd = entry.Balance.Sub(uncoveredPurchase)
+					uncoveredPurchase = sdk.ZeroInt()
+				}
+			}
+
+			// Make payout after purchased is fully covered.
+			if remainingUbd.IsPositive() {
+				if remainingUbd.GTE(payout) {
+					k.PayFromUnbonding(ctx, ubd, payout)
+					return
+				}
+				k.PayFromUnbonding(ctx, ubd, remainingUbd)
+				payoutFromUnbonding = payoutFromUnbonding.Sub(remainingUbd)
+			}
+		}
+	}
+
+	// Update collateral and withdraw.
+	provider.Collateral = provider.Collateral.Sub(payout)
+	provider.Withdrawing = provider.Withdrawing.Sub(payout)
+	k.SetProvider(ctx, provider.Address, provider)
+
+	// Update withdraw queue.
+	withdraws := k.GetWithdrawQueueTimeSlice(ctx, withdraw.CompletionTime)
+	for i := range withdraws {
+		if withdraws[i].Address.Equals(withdraw.Address) && withdraws[i].Amount.Equal(withdraw.Amount) {
+			withdraws[i].Amount = withdraws[i].Amount.Sub(payout)
+			break
+		}
+	}
+	k.SetWithdrawQueueTimeSlice(ctx, withdraw.CompletionTime, withdraws)
+}
+
+// GetSortedUnbondingDelegations gets unbonding delegations sorted by completion time from latest to earliest.
+func (k Keeper) GetSortedUnbondingDelegations(ctx sdk.Context, delAddr sdk.AccAddress) []staking.UnbondingDelegation {
+	ubds := k.sk.GetAllUnbondingDelegations(ctx, delAddr)
```

Comment:
Aren't the `ubd.Entries` already sorted here?

Reply:
`ubd.Entries` are sorted, but `ubds` are not.